### ISSUE TITLE
Multiprocessing with ImageDataGenerator (for augmentation) and  model.fit_generator

### ISSUE THREAD

**Issue by wittmarf** [keras-team/keras-preprocessing#145]
I need to use the model.fit_generator method with use_multiprocessing=True and workers>1 because I want to parallelize augmentation. One single CPU core is not able to deliver enough training data to the GPU to keep it busy.

The Keras ImageDataGenerator would be a perfect match for the model.fit_generator. But several people found out, that this causes problems because of lacking thread safety.

Is there a simple way to make those two work together? Not everyone working with Keras wants (or is able) to write a data generator for augmentation from scratch.

It would be greatly appreciated if you would give an example in the documentation.
In my case, the images that need to be augmented are in directories, one directory for each class.

Thanks!

**Comment by Dref360** [comment]
Do you have exemples where it's not safe? Using a keras.utils.Sequence, it should be safe.

**Comment by wittmarf** [comment]
Here is an example for the critical part of the program.
The problem is that in each epoch some images from the directories (before augmentation) are trained many times, while other images or not used at all. 

train_datagen = ImageDataGenerator(
    vertical_flip=True,
    rotation_range=20,
    zoom_range=0.1,
    horizontal_flip=True,
    preprocessing_function=normalize_RGB)

validation_datagen = ImageDataGenerator( ... )

train_generator = train_datagen.flow_from_directory(
    train_data_dir,
    target_size=(img_width, img_height),
    color_mode='rgb',
    batch_size=batch_size,
    class_mode='categorical')

validation_generator = validation_datagen.flow_from_directory( ...  )

# increase number of images found in directories with augmentation
nb_train_samples = 16*n_t
nb_validation_samples = 8*n_v

history = model.fit_generator(
    train_generator,
    steps_per_epoch=nb_train_samples//batch_size,
    epochs=epochs,
    validation_data=validation_generator,
    validation_steps=nb_validation_samples//batch_size,
    use_multiprocessing=True,
    workers=8)_

**Comment by Dref360** [comment]
Using the lates keras and keras_preprocessing, you shouldn't see this behaviour. We had a problem like this when keras_preprocessing splitted from Keras, but it should be resolved. 

Testing locally right now and I cannot reproduce.

**Comment by Dref360** [comment]
Do not ignore this.
Please update to the latest version of Keras. A bug was introduced when we splitted Keras-Preprocessing from Keras. It is now fixed.

**Comment by wittmarf** [comment]
Ok. Thank you!

**Comment by JasonKChow** [comment]
I'm using Keras 2.2.4 and preprocessing 1.0.5 and I still see the warning. I'm using ImageDataGenerator, flow_from_directory, and fit_generator as above. In addition, I am trying to fit a multi_gpu_model. How should I go about fixing this?  

Here is the relevant pieces of code:
```
trainGen = ImageDataGenerator(horizontal_flip = True,
                              channel_shift_range = 100)
trainData = trainGen.flow_from_directory(directory = dataDir + '/train',
                                         target_size = (224, 224),
                                         batch_size = 256,
                                         class_mode = 'categorical',
                                         seed = trainSeed)

valGen = ImageDataGenerator()
valData = valGen.flow_from_directory(directory = dataDir + '/sortedVal',
                                     target_size = (224, 224),
                                     batch_size = 256,
                                     class_mode = 'categorical',
                                     seed = valSeed)

vggNew.fit_generator(generator = trainData,
                     steps_per_epoch = 5000,
                     epochs = 100,
                     validation_data = valData,
                     validation_steps = 1000,
		     use_multiprocessing = True,
		     workers = workers,
		     callbacks = [reduceLR, earlyStop])
```

**Comment by wittmarf** [comment]
Hi Frederic,

I am still getting the  warning with Keras 2.2.4 and Keras-Preprocessing
1.0.9.

Greetings
Ralf

Am Fr., 25. Jan. 2019 um 15:03 Uhr schrieb Frédéric Branchaud-Charron <
notifications@github.com>:

**Comment by kroskal** [comment]
I also get this warning with Keras 2.2.4 and Keras-Preprocessing 1.0.5 :/

**Comment by dhruvmonga** [comment]
Hi,

I'm still having this problem with Keras 2.2.4 and Keras-Preprocessing 1.0.9 (latest versions that I can see). This may still be a bug. I can provide more information about my setup if needed!

**Comment by Dref360** [comment]
Keras has not made a new version with the fix in a while (many months actually).

To fix this issue: use keras master branch
```
pip uninstall keras
pip install git+https://github.com/keras-team/keras
```


I'm really sorry about that, I'll try to push for a new release.

**Comment by ahnchive** [comment]
Hi,
 
I'm trying to use ImageDataGenerator with multiprocessing and multiple works, but still got those warnings even with Keras master branch! Has anyone who had the same issues solved the problem?

Thanks

**Comment by AriMari90** [comment]
Hi
I have the same issue. I tried many Keras versions including the one from master branch. No success yet! Does anyone have updates about it?

**Comment by shrinivas22** [comment]
Hi, 
I am still facing this issue has there been any update? Any help would be great

**Comment by kylemcdonald** [comment]
I am also seeing this issue using `ImageDataGenerator` and `use_multiprocessing=True` with both Tensorflow 2.2.0 and Keras 2.4.3. The output is:

```
WARNING:tensorflow:Using a generator with `use_multiprocessing=True` and multiple workers may duplicate your data. Please consider using the `tf.data.Dataset`.
WARNING:tensorflow:multiprocessing can interact badly with TensorFlow, causing nondeterministic deadlocks. For high performance data pipelines tf.data is recommended.
```

Right now augmentation is the bottleneck in my training.

**Comment by Dref360** [comment]
As recommended by TF, using data pipeline might be the way to go if you don't use a custom Sequence object.

They support similar APIs as Keras-preprocessing: flow_from_directory --> dataset_from_directory for example.

**Comment by kylemcdonald** [comment]
@Dref360 thanks, i looked into that but it's a mess right now for anything beyond the basic tutorial. For example, if want to rebuild my augmentation with shear and rotation, I could use `keras.preprocessing.image.apply_affine_transform`, but this only operates on numpy arrays. Once `map` is called on a `tf.data.Dataset`, the objects being mapped are Tensors. There are some classes in [tensorflow_addons](https://www.tensorflow.org/addons/api_docs/python/tfa/image/transform_ops) that could help, but this leads to other problems.

**Comment by Dref360** [comment]
Sorry for the delay!
I should have time to look at this over the weekend. I was sure we fixed that issue a while ago.

Have you looked at PreprocessingLayers? In `tf.keras.layers.preprocessing` they should handle a wider range of use cases. When the size of the tensor doesnt change.

**Comment by Dref360** [comment]
Hello!
From the latest Keras release, it seems to work. We still get the warning `multiprocessing can interact badly with TensorFlow...`, but it still makes a `keras.utils.data_utils.OrderedEnqueuer` so it should be just like before.

Sorry it is a long example. 

```python
import keras
import numpy as np
import pytest

from PIL import Image

from keras_preprocessing.image import utils
from keras.preprocessing.image import ImageDataGenerator

@pytest.fixture(scope='module')
def all_test_images():
    img_w = img_h = 20
    rgb_images = []
    rgba_images = []
    gray_images = []
    for n in range(8):
        bias = np.random.rand(img_w, img_h, 1) * 64
        variance = np.random.rand(img_w, img_h, 1) * (255 - 64)
        imarray = np.random.rand(img_w, img_h, 3) * variance + bias
        im = Image.fromarray(imarray.astype('uint8')).convert('RGB')
        rgb_images.append(im)

        imarray = np.random.rand(img_w, img_h, 4) * variance + bias
        im = Image.fromarray(imarray.astype('uint8')).convert('RGBA')
        rgba_images.append(im)

        imarray = np.random.rand(img_w, img_h, 1) * variance + bias
        im = Image.fromarray(
            imarray.astype('uint8').squeeze()).convert('L')
        gray_images.append(im)

    return [rgb_images, rgba_images, gray_images]


@pytest.fixture(scope='module')
def image_data_generator():
    return ImageDataGenerator(
        featurewise_center=True,
        samplewise_center=True,
        featurewise_std_normalization=True,
        samplewise_std_normalization=True,
        zca_whitening=True,
        rotation_range=90.,
        width_shift_range=0.1,
        height_shift_range=0.1,
        shear_range=0.5,
        zoom_range=0.2,
        channel_shift_range=0.,
        brightness_range=(1, 5),
        fill_mode='nearest',
        cval=0.5,
        horizontal_flip=True,
        vertical_flip=True
    )


def test_numpy_array_iterator(image_data_generator, all_test_images, tmpdir):
    for test_images in all_test_images:
        img_list = []
        for im in test_images:
            img_list.append(utils.img_to_array(im)[None, ...])
        images = np.vstack(img_list)
        dsize = images.shape[0]
        channel_size = images.shape[-1]
        assert channel_size <= 4

        iterator = image_data_generator.flow(
            images,
            np.arange(images.shape[0]) % 3,
            shuffle=False,
            batch_size=3
        )

        img_inputs = keras.Input(shape=(20, 20, channel_size))
        x = keras.layers.Conv2D(32, 3,)(img_inputs)
        x = keras.layers.Flatten()(x)
        x = keras.layers.Dense(3)(x)
        model = keras.Model(inputs=img_inputs, outputs=x, name="mnist_model")
        model.compile(loss='mse')
        model.fit_generator(iterator, 10, epochs=1, use_multiprocessing=True, workers=4)

if __name__ == '__main__':
    pytest.main()
```
